### PR TITLE
JavaDependency: set JAVA_HOME and CFLAGS in superenv

### PR DESCRIPTION
**Pull request by xu-cheng**
Per https://github.com/Homebrew/homebrew/pull/36998#discussion_r25134850

### CODE REVIEW

File: Library/Homebrew/requirements.rb [hunk 1]
```diff
@@ -134,7 +146,8 @@ def initialize(tags)
   def java_version
     args = %w[/usr/libexec/java_home --failfast]
```
Comment (mikemcquaid) [inline Library/Homebrew/requirements.rb]: I know this is probably my bad code but it feels weird to have `/usr/libexec/java_home` as part of `args`; perhaps split that out to the `system` invocation.

File: Library/Homebrew/requirements.rb [hunk 2]
```diff
@@ -134,7 +146,8 @@ def initialize(tags)
...
     args << "--version" << "#{@version}+" if @version
-    quiet_system(*args)
+    @java_home = `#{args * " "}`.chomp
```
Comment (mikemcquaid) [inline Library/Homebrew/requirements.rb]: Mmm, backticks here allow you to run any arbritrary code by specificying it as the Java version e.g. `depends_on :java => "1.7+; rm -rf ~"`. I realise we probably have a bunch of other places where that's possible but would be nice to work around it if we can.

**Comment by mikemcquaid** [comment]
Looks good other than comment :+1:

**Comment by xu-cheng** [comment]
Updated.

File: Library/Homebrew/requirements.rb [hunk 3]
```diff
@@ -125,16 +125,28 @@ class JavaDependency < Requirement
...
+      ENV.append_to_cflags "-I#{java_home}/include/darwin"
+    else # Apple JVM
+      ENV.append_to_cflags "-I/System/Library/Frameworks/JavaVM.framework/Versions/Current/Headers/"
+    end
+  end
+
   def initialize(tags)
-    # ignore trailing +
-    @version = tags.shift.sub(/\+$/, "") if /(\d\.)+\d/ === tags.first
+    @version = tags.shift if /(\d\.)+\d/ === tags.first
```
Comment (xu-cheng) [inline Library/Homebrew/requirements.rb]: @mikemcquaid Why do we call `shift` here? It's not helpful for debug:
```
[1] brew(main)> JavaDependency.new ["1.8+"]
=> #<JavaDependency: "java" []>
```

Comment (mikemcquaid) [inline Library/Homebrew/requirements.rb]: So the arguments are not passed to the Requirement superclass; it's the way most other Requirements seem to behave.

Comment (xu-cheng) [inline Library/Homebrew/requirements.rb]: It also stop the arguments being passed to `inspect`.

Comment (jacknagel) [inline Library/Homebrew/requirements.rb]: Just define inspect for this class

Comment (xu-cheng) [inline Library/Homebrew/requirements.rb]: Ok.

However I quick scanned all the Requirements. The behaviors are inconsistent. [Some](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/requirements/maximum_macos_requirement.rb) doesn't remove the tags from superclass. [Other](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/requirements/language_module_dependency.rb) deliberately passes the arguments to superclass. And many of them who call `shift` fail to redefine `inspect`.

**Comment by xu-cheng** [comment]
Ready to merge?

**Comment by mikemcquaid** [comment]
:+1:

**Pull request closed without merging**